login.py

```diff
@@ -0,0 +1,219 @@
+
+        # if two factor auth detected
+        if error and error == "TFA":
+
+            # find the default tfa method
+            default_method, method_error = login_utils.get_default_method(response.text)
+
+            if not method_error:
+                # save current response url and payload for next request
+                payload = utils.make_payload(response.text)
+
+                if default_method == '1':
+                    query_params = utils.get_query_params(response.text)
+                    session.q_params = query_params
+
+                session.next_url = response.url
+                session.prev_payload = payload
+
+            # collect all available methods
+            methods, select_method_url, error, session = login_utils.select_alternate_method(
+                session, response.url)
+
+            if not error:
+                # save url to select methods, to make next request
+                session.select_method_url = select_method_url
+
+            # encode session as json
+            session = utils.serialize_session(session)
+
+            # if both default_method and available methods not fetched
+            if method_error and error:
+                resp.status = falcon.HTTP_500
+
+            # if available methods not fetched
+            elif error:
+                resp.status = falcon.HTTP_502
+                resp.body = json.dumps({'session': session, 'default_method': default_method})
+
+            # if default method not available
+            elif method_error:
+                resp.status = falcon.HTTP_503
+                resp.body = json.dumps({'session': session, 'methods': methods})
+
+            else:
+                # if both default method and available methods fetched
+                resp.status = falcon.HTTP_303
+                resp.body = json.dumps({'session': session, 'default_method': default_method,
+                                        'methods': methods})
+
+        elif error and error == "Connection Error":
+            resp.status = falcon.HTTP_504
+
+        elif error and error == "Parsing Error":
+            resp.status = falcon.HTTP_500
+
+        elif error and error == "Invalid credentials":
+            resp.status = falcon.HTTP_400
+
+        elif error and error == "captcha":
+            pass
+
+        else:
+            # encode session as json
+            session = jsonpickle.encode(session)
+
+            # if no two factor auth detected
+            resp.status = falcon.HTTP_200
+            resp.body = json.dumps({'session': session})
+
+
+@falcon.before(validate_request)
+class StepTwoLogin(object):
+    '''
+    Handles two factor authentication.
+    '''
+    def on_post(self, req, resp):
+        data = req.stream
+
+        method = data['method']
+        session = data['session']
+        session = utils.deserialize_session(session)
+
+        if method != 1:
+            otp = data['otp']
+            q_params = None
```

Comment:
not exactly, they are just some variables used in making request to google. will add comments here.